### PR TITLE
8266015: Implement AdapterHandlerLibrary lookup fast-path for common adapters

### PR DESCRIPTION
**Pull request by cl4es**
This patch refactors AdapterHandlerLibrary initialization so that we can initialize a handful of commonly used adapters early during bootstrap, and avoid taking the AdapterHandlerLibrary_lock when looking up these adapters.

Since the 5 most common adapters plus the abstract adapter constitutes roughly 60% of the method shapes loaded and linked on a Hello World, this means a relatively significant startup optimization (~2M insns on Hello World); most of the win is in lookup code that will be a significant part of the cost of class loading even when no adapters need to be generated.

This enhancement partially recuperates the regression reported in https://bugs.openjdk.java.net/browse/JDK-8265523

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266015](https://bugs.openjdk.java.net/browse/JDK-8266015): Implement AdapterHandlerLibrary lookup fast-path for common adapters

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3706/head:pull/3706` \
`$ git checkout pull/3706`

Update a local copy of the PR: \
`$ git checkout pull/3706` \
`$ git pull https://git.openjdk.java.net/jdk pull/3706/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3706`

View PR using the GUI difftool: \
`$ git pr show -t 3706`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3706.diff">https://git.openjdk.java.net/jdk/pull/3706.diff</a>

</details>

### CODE REVIEW

**Comment by cl4es** [comment]
To avoid making this code more complex, and enable splitting apart `get_adapter`, this patch simplifies the output for the develop flag `PrintAdapterHandlers` somewhat. Before this would arbitrarily print the signature of the method that happen to cause the adapter to be generated. Since adapters for methods with the same basic type arguments are shared it makes more sense to distill to BasicType strings.

Before:
```
AHE@0x00007efc503007a0: 0xbaabbaa0 i2c: 0x00007efc399484e0 c2i: 0x00007efc399485e2 c2iUV: 0x00007efc399485a8 c2iNCI: 0x00007efc3994861f
i2c argument handler #246 for: static ([Ljava/lang/Object;IILjava/util/Comparator;[Ljava/lang/Object;II)V 0xbaabbaa0 (614 bytes generated)
c2i argument handler starts at 0x7efc399485e2
```
After:
```
AHE@0x00007fc28c2f6850: 0x0baabbaa i2c: 0x00007fc27570d460 c2i: 0x00007fc27570d562 c2iUV: 0x00007fc27570d528 c2iNCI: 0x00007fc27570d59f
i2c argument handler #244 for: (LIILLII) 0x0baabbaa (614 bytes generated)
c2i argument handler starts at 0x7fc27570d562
```

File: src/hotspot/share/runtime/sharedRuntime.cpp [hunk 1]
```diff
@@ -2390,10 +2392,8 @@ class AdapterFingerPrint : public CHeapObj<mtCode> {
     int sig_index = 0;
     for (int index = 0; index < len; index++) {
       int value = 0;
-      for (int byte = 0; byte < _basic_types_per_int; byte++) {
-        int bt = ((sig_index < total_args_passed)
-                  ? adapter_encoding(sig_bt[sig_index++])
-                  : 0);
+      for (int byte = 0; sig_index < total_args_passed && byte < _basic_types_per_int; byte++) {
+        int bt = adapter_encoding(sig_bt[sig_index++]);
```
Comment (iklam) [inline src/hotspot/share/runtime/sharedRuntime.cpp:2396]: Why is the above changed? Is it for performance? It seems unrelated to this PR.

File: src/hotspot/share/runtime/sharedRuntime.cpp [hunk 2]
```diff
@@ -2637,150 +2734,178 @@ AdapterHandlerEntry* AdapterHandlerLibrary::new_entry(AdapterFingerPrint* finger
...
-      // CodeCache is full, disable compilation
-      // Ought to log this but compile log is only per compile thread
-      // and we're some non descript Java thread.
-      return NULL; // Out of CodeCache space
-    }
-    entry->relocate(new_adapter->content_begin());
+  new_adapter = AdapterBlob::create(&buffer);
+  NOT_PRODUCT(int insts_size = buffer.insts_size());
+  if (new_adapter == NULL) {
+    fatal("No memory to create adapter");
```
Comment (iklam) [inline src/hotspot/share/runtime/sharedRuntime.cpp:2883]: This used to be `return NULL;`, and the JVM is supposed to continue execution (but the linking of the current class will get a VirtualMachineError. I think we should keep the old behavior.

**Review by iklam — CHANGES_REQUESTED** [review]
Looks good overall, but I think we should avoid changing the `return NULL` to `fatal()`.

Also note to other reviewers: when looking at the "Files changed", add `&w=1` to the URL so GitHub will ignore whitespace changes.

File: src/hotspot/share/runtime/sharedRuntime.cpp (hunk 1 above)
Comment (cl4es) [reply src/hotspot/share/runtime/sharedRuntime.cpp:2396]: Performance (fingerprints are calculated on every lookup) but also for simplifying the fingerprint iterator in as_basic_args_string: Once we hit total_args_passed, the previous algorithm would keep shifting and adding zero bits until the int is "filled". Now I stop when the last argument has been processed, which meant that iterating becomes simpler. Since the hash function will xor the last value (which is the only one that can have zeros) as a last step, it should also improve how well trailing arguments shuffles low bits.

I'm adding some comments, and also fixed a bug in as_basic_args_string when > 8 arguments and changed as_string to not include leading zeroes of the last in (0xbbbbbabb0000000b -> 0xbbbbbabbb)

File: src/hotspot/share/runtime/sharedRuntime.cpp (hunk 2 above)
Comment (cl4es) [reply src/hotspot/share/runtime/sharedRuntime.cpp]: Yes, an unintentional result of me breaking the code flow apart and then merging it back together.

**Comment by cl4es** [comment]
@iklam asked me offline to experiment using a concurrent hash table keyed on fingerprint to create a quick filter. I've done a few prototypes and verified that a concurrent hash table alone[1] does perform roughly as well as this patch on a few sampled applications, and that combining the current patch with a concurrent hash table comes out slightly ahead of both[2]. But this comes at a cost of an additional data structure that might grow to non-trivial size.

After analyzing the most significant part of the difference comes not from a reduction in `Mutex` overhead (~400k insts) but from a reduction in `SignatureStream` parsing (~1.2M insns). Using a `SignatureIterator` that takes the method fingerprint into account gets us ahead of the combined `ConcurrentHashTable` experiment, without needing to add another data structure. 

For a follow-up to this it might make sense to re-implement `AdapterHandlerLibrary::_adapters` as a `ConcurrentHashTable`  rather than a `BasicHashtable`.

[1] https://github.com/openjdk/jdk/compare/master...cl4es:adapter_cht_only?expand=1
[2] https://github.com/openjdk/jdk/compare/master...cl4es:adapter_cht_combined?expand=1

**Review by iklam — APPROVED** [review]
The latest version LGTM. Thanks for also testing out the alternative optimizations.

**Review by coleenp — APPROVED** [review]
This looks good to me!

**Comment by cl4es** [comment]
@iklam @coleenp - thank you for reviewing!

/integrate
